### DISCUSSION TITLE
Floating point exception on output runoff & base flow within HyMAP2

### DISCUSSION THREAD

**Discussion by jnattala** [NASA-LIS/LISF#1510] [Q&A]
Hello,

This is regarding an overflow error on output runoff & base flow within HyMAP2.

Undefined values passed to HYMAP2_model_core seem to rely on a type conversion to real, of integer components 'imis', of HYMAP2_routing_struc array variables of derived type HYMAP2_routing_dec (i.e. actual arg. real(HYMAP2_routing_struc(:)%imis) within call to HYMAP2_model, from HYMAP2_routing_run).

For my configuration, HYMAP2_routing_struc(:)%imis = -9999 & HYMAP2_routing_struc(n)%evapflag is zero. This causes a overflow on output runoff & baseflow (runoff1 & basflw1), when dwiflag is zero, for a river sequence.

Below is a portion of a slightly modified version of lislog.0000 output, from a call to HYMAP2_grid2vector, from HYMAP2_routing_run (for imported surface runoff & baseflow), for some portions of LIS cycle times. The last six columns within modified log output from HYMAP2_grid2vector, show real(LIS_rc%udef), HYMAP2_routing_struc(n)%evapflag, ix, iy & nz. These are followed by corresponding modified log output, from HYMAP2_model_core.

[INFO] LIS cycle time: 07/30/2023 23:45:00
  [JN]   HYMAP2_grid2vector   LIS cycle time: 07/30/2023 23:45:00    306321  surface_runoff(i,:):     0.00000E+00       -9999.00    0  729  799   1
  [JN]   HYMAP2_grid2vector   LIS cycle time: 07/30/2023 23:45:00    306321  baseflow(i,:)      :     0.17540E-09       -9999.00    0  729  799   1
  [JN]   HYMAP2_model_core    LIS cycle time: 07/30/2023 23:45:00    306321  basflw0(ic):     0.17540E-09  runoff0(ic):     0.00000E+00   grarea(ic):    208000000.00000

[INFO] LIS cycle time: 07/31/2023 23:30:00
  [JN]   HYMAP2_grid2vector   LIS cycle time: 07/31/2023 23:30:00    306321  surface_runoff(i,:):     0.99692E+37       -9999.00    0  729  799    1
  [JN]   HYMAP2_grid2vector   LIS cycle time: 07/31/2023 23:30:00    306321  baseflow(i,:)      :     0.99692E+37       -9999.00    0  729  799    1
  [JN]   HYMAP2_model_core    LIS cycle time: 07/31/2023 23:30:00    306321  basflw0(ic):     0.99692E+37  runoff0(ic):     0.99692E+37   grarea(ic):    208000000.00000

When fill values on input runoff (runoff0) & input baseflow (basflw0) are not identical to real(LIS_rc%udef), as seen above for LIS cycle time 07/31/2023 23:30:00, there is a FPE on output runoff (runoff1) & baseflow (basflw1) (copied below).

Unsure if this is a known issue, if not, I have implemented a minor fix based on an absolute difference magnitude comparison, and can furnish it if needed. Please let me know if it would be useful.

thanks
Jyothi

FPE:
---
forrtl: error (75): floating point exception
LIS_I19_SL12       0000000000499BA1  hymap2_model_core         140  HYMAP2_model_core.F90
LIS_I19_SL12       0000000000498263  hymap2_model_             200  HYMAP2_model.F90
LIS_I19_SL12       00000000004BD466  hymap2_routing_ru         676  HYMAP2_routing_run.F90
LIS_I19_SL12       000000000066083F  lis_routingmod_mp         131  LIS_routingMod.F90
LIS_I19_SL12       0000000000F911D3  retrospective_run         206  retrospective_runMod.F90
LIS_I19_SL12       0000000000C01044  MAIN__                     68  lisdrv.F90

**Comment by jvgeiger** [comment]
Thanks for this issue.

@agetiran, would you please review this?

**Reply by agetiran** [reply]
I had a meeting with @jnattala earlier this week to talk about this issue. It looks like the floating point exception is occurring due to questionable values (i.e., 0.99692E+37) found in Qs and Qsb fields used as inputs in HyMAP. The problem seems to be related to either Noah-MP or LIS output file writing, not HyMAP.

**Comment by jvgeiger** [comment]
Thanks.  I will close this discussion.
